File: visitor-comparison/src/main/java/com/edwyn/demo/visitor/java25/review/ValidationVisitor.java
package com.edwyn.demo.visitor.java25.review;

import com.edwyn.demo.visitor.java25.order.DigitalProduct;
import com.edwyn.demo.visitor.java25.order.Discount;
import com.edwyn.demo.visitor.java25.order.OrderItemVisitor;
import com.edwyn.demo.visitor.java25.order.PhysicalProduct;
import com.edwyn.demo.visitor.java25.order.ServiceLine;

import java.util.ArrayList;
import java.util.List;

public final class ValidationVisitor implements OrderItemVisitor<List<String>> {
    @Override
    public List<String> visitPhysicalProduct(PhysicalProduct item) {
        var issues = new ArrayList<String>();
        var label = label(item.name(), "Physical product");
        requireName(item.name(), "Physical product", issues);
        requirePositive(item.unitPriceCents(), label + " unit price", issues);
        requirePositive(item.quantity(), label + " quantity", issues);
        requirePositive(item.weightGrams(), label + " weight", issues);
        return List.copyOf(issues);
    }

    @Override
    public List<String> visitDigitalProduct(DigitalProduct item) {
        var issues = new ArrayList<String>();
        var label = label(item.name(), "Digital product");
        requireName(item.name(), "Digital product", issues);
        requirePositive(item.unitPriceCents(), label + " unit price", issues);
        requirePositive(item.licenseCount(), label + " license count", issues);
        return List.copyOf(issues);
    }

    @Override
    public List<String> visitServiceLine(ServiceLine item) {
        var issues = new ArrayList<String>();
        var label = label(item.name(), "Service");
        requireName(item.name(), "Service", issues);
        requirePositive(item.hourlyRateCents(), label + " hourly rate", issues);
        requirePositive(item.hours(), label + " hours", issues);
        if (item.hours() > 40) {
            issues.add(label + " cannot exceed 40 hours");
        }
        return List.copyOf(issues);
    }

    @Override
    public List<String> visitDiscount(Discount item) {
        var issues = new ArrayList<String>();
        requireName(item.code(), "Discount code", issues);
        requirePositive(item.amountCents(), item.code() + " discount amount", issues);
        return List.copyOf(issues);
    }

    private static String label(String value, String fallback) {
        return value == null || value.isBlank() ? fallback : value;
    }

    private static void requireName(String value, String label, List<String> issues) {
        if (value == null || value.isBlank()) {
            issues.add(label + " name is required");
        }
    }

    private static void requirePositive(int value, String label, List<String> issues) {
        if (value <= 0) {
            issues.add(label + " must be positive");
        }
    }
}

File: visitor-comparison/src/main/kotlin/com/edwyn/demo/visitor/kotlin/review/ValidationVisitor.kt
package com.edwyn.demo.visitor.kotlin.review

import com.edwyn.demo.visitor.kotlin.order.DigitalProduct
import com.edwyn.demo.visitor.kotlin.order.Discount
import com.edwyn.demo.visitor.kotlin.order.OrderItem
import com.edwyn.demo.visitor.kotlin.order.PhysicalProduct
import com.edwyn.demo.visitor.kotlin.order.ServiceLine

class ValidationVisitor {
    fun visit(item: OrderItem): List<String> =
        when (item) {
            is PhysicalProduct -> validatePhysicalProduct(item)
            is DigitalProduct -> validateDigitalProduct(item)
            is ServiceLine -> validateServiceLine(item)
            is Discount -> validateDiscount(item)
        }

    private fun validatePhysicalProduct(item: PhysicalProduct): List<String> = buildList {
        val label = item.name.ifBlank { "Physical product" }
        validateName(item.name, "Physical product")
        validatePositive(item.unitPriceCents, "$label unit price")
        validatePositive(item.quantity, "$label quantity")
        validatePositive(item.weightGrams, "$label weight")
    }

    private fun validateDigitalProduct(item: DigitalProduct): List<String> = buildList {
        val label = item.name.ifBlank { "Digital product" }
        validateName(item.name, "Digital product")
        validatePositive(item.unitPriceCents, "$label unit price")
        validatePositive(item.licenseCount, "$label license count")
    }

    private fun validateServiceLine(item: ServiceLine): List<String> = buildList {
        val label = item.name.ifBlank { "Service" }
        validateName(item.name, "Service")
        validatePositive(item.hourlyRateCents, "$label hourly rate")
        validatePositive(item.hours, "$label hours")
        validateMaxHours(item.hours, label, maxHours = 40)
    }

    private fun validateDiscount(item: Discount): List<String> = buildList {
        validateName(item.code, "Discount code")
        validatePositive(item.amountCents, "${item.code} discount amount")
    }

    private fun MutableList<String>.validateName(value: String, label: String) {
        if (value.isBlank()) {
            add("$label name is required")
        }
    }

    private fun MutableList<String>.validatePositive(value: Int, label: String) {
        if (value <= 0) {
            add("$label must be positive")
        }
    }

    private fun MutableList<String>.validateMaxHours(hours: Int, label: String, maxHours: Int) {
        if (hours > maxHours) {
            add("$label cannot exceed $maxHours hours")
        }
    }
}
